### PR TITLE
Include AuthenticationCompletionException messages in HTTP response in devmode

### LINKED ISSUE #41890: Return OIDC code flow error_code and error_description directly to the user if no error handler is available
### Description

If the OIDC code flow fails, then a user can point to to an error handler endpoint with `quarkus.oidc.authentication.error-path`. Error handler can provide a nicely formatted page, and suggest a user to retry by following some link, etc.

If the error handler is not available, 401 is returned. But it would be better return the error code and description directly to the user as a simple text message. Users can see them in the browser URL but it is not easy

### Implementation ideas

Return a text message including the error code and description if no error handler is available

### PR DESCRIPTION
**Pull request by sberyozkin**
Fixes #41890.

It really fixes a bug rather than a planned enhancement.

Sometimes, when the user is redirected to the OIDC provider and something goes wrong at the OIDC provider's side, an error code and optional error description is returned. 

 Awhile back, @FroMage asked about an option to customize such error responses so now users can register error handlers and refer to them with `quarkus.oidc.authentication.error-path` property.

But if they haven't registered it, they will get a rather obscure warning and 401 in the browser without any further clues.

@maxandersen was concerned about it the other day, but also @Sanne and indeed @Fromage have also been concerned.

So first of all, this PR improves a log message reported when the user is redirected back to Quarkus with the error code.
And it goes over all/most of `AuthenticationCompletionException` cases,  initializes `AuthenticationCompletionException` with the error message, and allows this error message be reported back to the user, but in devmode only, to avoid accidentally leaking too many details.

### CODE REVIEW

File: extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java [hunk 1]
```diff
@@ -240,13 +241,19 @@ public Uni<SecurityIdentity> apply(TenantConfigContext tenantContext) {
...
                         });
             } else {
-                LOG.error(
-                        "Authentication has failed but no error handler is found, completing the code flow with HTTP status 401");
-                return Uni.createFrom().failure(new AuthenticationCompletionException());
+                final String errorMessage = """
+                        Authorization code flow has failed, error code: %s, error description: %s.
+                        Error handler path is not configured. Have a public JAX-RS resource listening
+                        on a path such as '/error' and point to it with 'quarkus.oidc.authentication.error-path=/error'.
+                        Completing the flow with HTTP status 401.
```
Comment (michalvavrik) [inline extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java:248]: It's quite big, logging is expensive and this is logger on error level. If I am correct, this will only be logged for when setup is wrong either on IDM or Quarkus or client side (or someone is using manual requests etc.). In that case it makes sense to log error code and description. But the error handler seems like useful information in DEV mode only. Valid token is required and incoming request doesn't have it, so it must be rejected. The error handler sounds like something regular user won't do, they cannot just change incoming credentials.

In short:

- do I read it right that this won't ever be logged for normal situations like when access token expired during request to UserInfo endpoint?
- is the mention about error handler useful to small amount of cases or do you really expect users to use it? I was alarmed at mention that Stephane requested it, because it would suggest you didn't hear often from users that they need it.

File: extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java [hunk 2]
```diff
@@ -175,9 +175,10 @@ private Uni<SecurityIdentity> processRedirectFromOidc(RoutingContext context, Oi
         String[] parsedStateCookieValue = COOKIE_PATTERN.split(stateCookie.getValue());
         OidcUtils.removeCookie(context, oidcTenantConfig, stateCookie.getName());
         if (!parsedStateCookieValue[0].equals(stateQueryParam.get(0))) {
-            LOG.debug("State cookie value does not match the state query parameter value, "
-                    + "completing the code flow with HTTP status 401");
-            return Uni.createFrom().failure(new AuthenticationCompletionException());
+            final String error = "State cookie value does not match the state query parameter value, "
```
Comment (michalvavrik) [inline extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java:178]: I'd like to understand if this can happen in normal situation.

For example I open all the new links in new tabs because I simply don't trust SPAs they are well written and I prefer to keep the tab I am working with intact. But parallel code flows are not supported. Can this happen that state cookie value doesn't match the state query parameter because I work in multiple tabs? Because I don't think it would be useful for administrators to read this particualr scenario.

**Review by michalvavrik — COMMENTED** [review]
Hey,

first of all, I think returning error message in DEV mode is right step and it will help debugging. As far as `AuthenticationCompletionException` changes are concerned LGTM.

As you are changing logging level to error and some error messages are verbose, I'd like to check all the logged messages are actionable (provides information on which normally admin will act, but does not describe regular situation). I'll comment on individual changes.

File: extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java (hunk 1 above)
Comment (sberyozkin) [reply extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java:248]: @michalvavrik 

>It's quite big, logging is expensive and this is logger on error level. 

This covers an exception path  where the user is redirected back from Keycloak or other provider to Quarkus, but the user has failed to authenticate at the provider side for whatever reason. I wanted to have a very clear message how to intercept such flows (where a custom handler for ex returns a nice HTML page to the user explaining what happened). 
Compare it with the original message. It did not help Max in the slightest when he asked me what was going, what is the error handler, etc. I'm not worried at all abut the cost of logging in this case -:)

>do I read it right that this won't ever be logged for normal situations like when access token expired during request to UserInfo endpoint?

This specific one handles a moment where the authorization code flow is in progress and only the first part of it has completed, in this case, instead of returning `code` to Quarkus for it to be exchanged for tokens, the user is redirected with an error code.

>is the mention about error handler useful to small amount of cases or do you really expect users to use it? I was alarmed at mention that Stephane requested it, because it would suggest you didn't hear often from users that they need it.

I'm not sure what do you mean. Such errors are not happening often but it does not mean the handler is not needed. Steph asked for it a couple of years back, it was useful to have it added. But users, if they haven't done it, and start wondering in prod, why exactly the screen is empty now, they will see in the log what they need to do to make things better...

File: extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java (hunk 2 above)
Comment (sberyozkin) [reply extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java:178]: @michalvavrik This is actually covers a case where a parallel tab login is now allowed...

But to be honest, I'm not sure why we are looking at individual log messages. The idea of the PR is to get 401 related errors returned to users in devmode.

File: extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java (hunk 1 above)
Comment (sberyozkin) [reply extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java:248]: @michalvavrik I had a similar comment from users re the case where the session cookie is > 4K, initially I logged something short, but there was an issue from users asking to make it clear in the logs what to do...

Comment (michalvavrik) [reply extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java:248]: I won't answer individual comments. I think you have explained your position. This is not attack on Stephan, it was my worry that you document edge feature in the logging message while users normally shouldn't use these loggers. I don't know what most users would do in that error handler, they cannot change that credentials are wrong, they can alter error message but then, they are giving away information in production.

Thank you for the explanation, let's leave it there.

File: extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java (hunk 2 above)
Comment (michalvavrik) [reply extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java:178]: > But to be honest, I'm not sure why we are looking at individual log messages. The idea of the PR is to get 401 related errors returned to users in devmode.

You have changed logging level from "enabled when I am debugging" to enabled always (because everyone is logging errors). Our users pay for these logs and logging is slowing down things, so I am trying to figure if they actually will find these messages useful in production. You said yes and I believe you.

**Review by michalvavrik — APPROVED** [review]

File: extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java (hunk 1 above)
Comment (sberyozkin) [reply extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java:248]: @michalvavrik Yeah, I know you meant to clarify things, np at all.

These error code and description, if any, they go via the browser, so these are standard original OAuth2 properties, so they are not sensitive properties as such. 
Now, the error handler can choose to return HTML page to the user, explaining with the error code alone that this or that part of the flow did not work out. 
Or it can just rethrow 401 but get hold of this error code and description and do something with it, like sending an alert or do something else...
Without the error handler - Quarkus app won't know what happened...

File: extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java (hunk 2 above)
Comment (sberyozkin) [reply extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java:178]: @michalvavrik I see what I mean, it is just that 401 deserves a more visible kind of message, we have plenty of debug messages which will stay at the debug level.

In general, I'd like to make sure that when the user gets 401, for example, in prod, the teams don't scratch their heads why the user got 401 in that case, we are keeping getting messages that it is not easy to find out why the failure occurred

Hope it makes sense

**Comment by sberyozkin** [comment]
I'll ask for more reviews tomorrow, thanks @michalvavrik

File: extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java (hunk 1 above)
Comment (michalvavrik) [reply extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java:248]: > ... Without the error handler - Quarkus app won't know what happened...

I think being able to intercept error and propagate it in own manner makes sense. Or to produce standardized error code.

@sberyozkin I cannot close comment threads in your PRs (rights) :-) please consider it closed and thank you for the explanation.

File: extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java (hunk 2 above)
Comment (michalvavrik) [reply extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java:178]: > In general, I'd like to make sure that when the user gets 401, for example, in prod, the teams don't scratch their heads why the user got 401 in that case, we are keeping getting messages that it is not easy to find out why the failure occurred

Yes, I think you confirm what I wanted to know. That when 401 occurrences, they can use logged information to fix issue, not just to be informed that something regular happened. I tried to figure from flow of the code that leads to these messages, but it's not always straightforward.

Thanks, please consider this thread closed.

File: extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java (hunk 1 above)
Comment (sberyozkin) [reply extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java:248]: @michalvavrik Np at all, thanks too.

**Review by FroMage — APPROVED** [review]
Looks like a great improvement, thanks!

**Comment by maxandersen** [comment]
looks good - would be great if you showed a before and after example :)

p.s. I don't have a easy setup atm to try it out atm but let me know if need to and i can try redo it.

**Comment by sberyozkin** [comment]
There is also an optional `error_uri` there, but I'd rather avoid suggesting users go to another website to get even a more clearer description than what `error_description` can provide; using local error handlers to return nice HTML pages if it is what is required would be safer
